crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs

```diff
@@ -0,0 +1,171 @@
+                            );
+                        }
+                        // callee belongs to module, scope switches to module and class becomes caller
+                        ExprKind::Name {
+                            id: callee_name, ..
+                        } => {
+                            if let Some(scope) = checker.ctx.grandparent_scope() {
+                                return func_defined_above(
+                                    checker,
+                                    func,
+                                    scope,
+                                    callee_name,
+                                    class_name,
+                                );
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                Scope {
+                    // caller is defined in module
+                    kind: ScopeKind::Module,
+                    ..
+                } => {
+                    // callee belongs to module
+                    if let ExprKind::Name {
+                        id: callee_name, ..
+                    } = node
+                    {
+                        return func_defined_above(checker, func, scope, callee_name, caller_name);
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    None
+}
+
+// check if callee function that is called from caller is defined above caller
+fn func_defined_above(
+    checker: &Checker,
+    func: &Expr,
+    scope: &Scope,
+    callee_name: &String,
+    caller_name: &str,
+) -> Option<Diagnostic> {
+    if let (Some(callee_binding_idx), Some(caller_binding_idx)) =
+        (scope.get(&callee_name[..]), scope.get(caller_name))
+    {
+        if caller_binding_idx > callee_binding_idx {
+            if let Binding {
+                source:
+                    Some(RefEquality(Located {
+                        node: StmtKind::FunctionDef { body, .. },
+                        ..
+                    })),
+                kind,
+                range,
+                ..
+            } = &checker.ctx.bindings[*callee_binding_idx]
+            {
+                if matches!(kind, BindingKind::FunctionDefinition)
+                    & !mutual_recursion(body, &caller_name.to_string())
+                {
+                    // report a violation if binding is a FunctionDefinition or no mutual recursion occurs
+                    let kind = FuncDefinedAbove {
+                        function_name: callee_name.to_string(),
+                        function_def_line: range.location.row(),
+                    };
+                    return Some(Diagnostic::new(kind, Range::from(func)));
+                }
+            }
+        }
+    }
+    None
+}
+
+// check if caller function is also called from callee function (mutual recursion)
+fn mutual_recursion(body: &[Located<StmtKind>], caller_name: &String) -> bool {
```

Comment:
Nit: use `Stmt` in lieu of `Located<StmtKind>`, and `&str` in lieu of `&String`, if possible?

Reply:
Fixed